Sync error filling logs

**Describe the bug**
A database key violation is causing a lot of log entries.

**Logs**

```
2021-08-10 01:02:29,778 WARN sync.util :: Error syncing Fields for Table 'Table 723 'public.customer_table_name''
java.sql.BatchUpdateException: Batch entry 0 UPDATE "metabase_field" SET "database_type" = 'float8', "base_type" = 'type/Float', "database_position" = 24, "name" = 'Foo', "updated_at" = now() WHERE "id" = 64650 was aborted: ERROR: duplicate key value violates unique constraint "idx_uniq_field_table_id_parent_id_name_2col"
  Detail: Key (table_id, name)=(723, Ballotbox) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:169)
	at org.postgresql.core.ResultHandlerDelegate.handleError(ResultHandlerDelegate.java:52)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2286)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:521)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:870)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:893)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1644)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2544)
	at clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:598)
	at clojure.java.jdbc$execute_batch.invoke(jdbc.clj:591)
	at clojure.java.jdbc$db_do_execute_prepared_statement$fn__8648.invoke(jdbc.clj:1057)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:807)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:789)
	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1056)
	at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1042)
	at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1080)
	at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1060)
	at clojure.java.jdbc$execute_BANG_$execute_helper__8716.invoke(jdbc.clj:1464)
	at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1468)
	at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1435)
	at toucan.db$execute_BANG_.invokeStatic(db.clj:427)
	at toucan.db$execute_BANG_.doInvoke(db.clj:422)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at toucan.db$update_BANG_.invokeStatic(db.clj:482)
	at toucan.db$update_BANG_.invoke(db.clj:471)
	at toucan.db$update_BANG_.invokeStatic(db.clj:491)
	at toucan.db$update_BANG_.invoke(db.clj:471)
	at metabase.sync.sync_metadata.fields.sync_metadata$fn__55819$update_field_metadata_if_needed_BANG___55824$fn__55825.invoke(sync_metadata.clj:96)
	at metabase.sync.sync_metadata.fields.sync_metadata$fn__55819$update_field_metadata_if_needed_BANG___55824.invoke(sync_metadata.clj:17)
	at metabase.sync.sync_metadata.fields.sync_metadata$fn__55881$update_metadata_BANG___55886$fn__55887$iter__55888__55892$fn__55893.invoke(sync_metadata.clj:121)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$filter$fn__5893.invoke(core.clj:2809)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.Cons.next(Cons.java:39)
	at clojure.lang.RT.next(RT.java:713)
	at clojure.core$next__5386.invokeStatic(core.clj:64)
	at clojure.core.protocols$fn__8159.invokeStatic(protocols.clj:169)
	at clojure.core.protocols$fn__8159.invoke(protocols.clj:124)
	at clojure.core.protocols$fn__8114$G__8109__8123.invoke(protocols.clj:19)
	at clojure.core.protocols$seq_reduce.invokeStatic(protocols.clj:27)
	at clojure.core.protocols$fn__8146.invokeStatic(protocols.clj:75)
	at clojure.core.protocols$fn__8146.invoke(protocols.clj:75)
	at clojure.core.protocols$fn__8088$G__8083__8101.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6824)
	at clojure.core$reduce.invoke(core.clj:6810)
	at metabase.sync.util$sum_for_STAR_.invokeStatic(util.clj:485)
	at metabase.sync.util$sum_for_STAR_.invoke(util.clj:482)
	at metabase.sync.sync_metadata.fields.sync_metadata$fn__55881$update_metadata_BANG___55886$fn__55887.invoke(sync_metadata.clj:117)
	at metabase.sync.sync_metadata.fields.sync_metadata$fn__55881$update_metadata_BANG___55886.invoke(sync_metadata.clj:111)
	at metabase.sync.sync_metadata.fields$fn__55923$sync_and_update_BANG___55928$fn__55929.invoke(fields.clj:63)
	at metabase.sync.sync_metadata.fields$fn__55923$sync_and_update_BANG___55928.invoke(fields.clj:55)
	at metabase.sync.sync_metadata.fields$fn__55951$sync_fields_for_table_BANG___55960$fn__55963$fn__55964.invoke(fields.clj:74)
	at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:156)
	at metabase.sync.util$do_with_error_handling.invoke(util.clj:149)
	at metabase.sync.sync_metadata.fields$fn__55951$sync_fields_for_table_BANG___55960$fn__55963.invoke(fields.clj:71)
	at metabase.sync.sync_metadata.fields$fn__55951$sync_fields_for_table_BANG___55960.invoke(fields.clj:65)
	at clojure.core$partial$fn__5839.invoke(core.clj:2624)
	at clojure.core$map$fn__5866.invoke(core.clj:2753)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.RT.seq(RT.java:535)
	at clojure.core$seq__5402.invokeStatic(core.clj:137)
	at clojure.core$filter$fn__5893.invoke(core.clj:2809)
	at clojure.lang.LazySeq.sval(LazySeq.java:42)
	at clojure.lang.LazySeq.seq(LazySeq.java:51)
	at clojure.lang.ChunkedCons.chunkedNext(ChunkedCons.java:59)
	at clojure.core$chunk_next.invokeStatic(core.clj:708)
	at clojure.core$reduce1.invokeStatic(core.clj:942)
	at clojure.core$reduce1.invokeStatic(core.clj:934)
	at clojure.core$merge_with.invokeStatic(core.clj:3059)
	at clojure.core$merge_with.doInvoke(core.clj:3051)
	at clojure.lang.RestFn.applyTo(RestFn.java:139)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.sync.sync_metadata.fields$fn__55989$sync_fields_BANG___55994$fn__55995.invoke(fields.clj:85)
	at metabase.sync.sync_metadata.fields$fn__55989$sync_fields_BANG___55994.invoke(fields.clj:77)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.sync.util$fn__40388$run_step_with_metadata__40393$fn__40397$fn__40399.doInvoke(util.clj:360)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:99)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:93)
	at metabase.sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:116)
	at metabase.sync.util$with_start_and_finish_debug_logging.invoke(util.clj:113)
	at metabase.sync.util$fn__40388$run_step_with_metadata__40393$fn__40397.invoke(util.clj:355)
	at metabase.sync.util$fn__40388$run_step_with_metadata__40393.invoke(util.clj:350)
	at metabase.sync.util$fn__40579$run_sync_operation__40584$fn__40585$fn__40593.invoke(util.clj:451)
	at metabase.sync.util$fn__40579$run_sync_operation__40584$fn__40585.invoke(util.clj:449)
	at metabase.sync.util$fn__40579$run_sync_operation__40584.invoke(util.clj:443)
	at metabase.sync.sync_metadata$fn__56661$sync_db_metadata_BANG___56666$fn__56667$fn__56668.invoke(sync_metadata.clj:49)
	at metabase.sync.util$do_with_error_handling.invokeStatic(util.clj:156)
	at metabase.sync.util$do_with_error_handling.invoke(util.clj:149)
	at clojure.core$partial$fn__5841.invoke(core.clj:2630)
	at metabase.driver$fn__29807.invokeStatic(driver.clj:527)
	at metabase.driver$fn__29807.invoke(driver.clj:527)
	at clojure.lang.MultiFn.invoke(MultiFn.java:239)
	at metabase.sync.util$sync_in_context$fn__40300.invoke(util.clj:135)
	at metabase.sync.util$with_db_logging_disabled$fn__40297.invoke(util.clj:126)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:99)
	at metabase.sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:93)
	at metabase.sync.util$with_start_and_finish_logging$fn__40286.invoke(util.clj:111)
	at metabase.sync.util$with_sync_events$fn__40281.invoke(util.clj:85)
	at metabase.sync.util$with_duplicate_ops_prevented$fn__40272.invoke(util.clj:64)
	at metabase.sync.util$do_sync_operation.invokeStatic(util.clj:177)
	at metabase.sync.util$do_sync_operation.invoke(util.clj:174)
	at metabase.sync.sync_metadata$fn__56661$sync_db_metadata_BANG___56666$fn__56667.invoke(sync_metadata.clj:48)
	at metabase.sync.sync_metadata$fn__56661$sync_db_metadata_BANG___56666.invoke(sync_metadata.clj:45)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invokeStatic(sync_databases.clj:67)
	at metabase.task.sync_databases$sync_and_analyze_database_BANG_.invoke(sync_databases.clj:58)
	at metabase.task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:75)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:557)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "idx_uniq_field_table_id_parent_id_name_2col"
  Detail: Key (table_id, name)=(723, Ballotbox) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285)
	... 117 more
```

**Information about your Metabase Installation:**

Metabase 1.40.1 in MB Cloud

--- a/src/metabase/sync/sync_metadata/fields/common.clj
+++ b/src/metabase/sync/sync_metadata/fields/common.clj
@@ -60,12 +60,28 @@
 
 (mu/defn matching-field-metadata :- [:maybe TableMetadataFieldWithOptionalID]
   "Find Metadata that matches `field-metadata` from a set of `other-metadata`, if any exists. Useful for finding the
-  corresponding Metabase Field for field metadata from the DB, or vice versa."
+  corresponding Metabase Field for field metadata from the DB, or vice versa. Will prefer exact matches."
   [field-metadata :- TableMetadataFieldWithOptionalID
    other-metadata :- [:set TableMetadataFieldWithOptionalID]]
-  (some
-   (fn [other-field-metadata]
-     (when (= (canonical-name field-metadata)
-              (canonical-name other-field-metadata))
-       other-field-metadata))
-   other-metadata))
+  (let [matches (keep
+                  (fn [other-field-metadata]
+                    (when (= (canonical-name field-metadata)
+                             (canonical-name other-field-metadata))
+                      other-field-metadata))
+                  other-metadata)
+        num-matches (count matches)]
+    (cond
+      (zero? num-matches)
+      nil
+
+      (= 1 num-matches)
+      (first matches)
+
+      :else
+      (or
+        (some (fn [match]
+                (when (= (:name field-metadata) (:name match))
+                  match))
+              matches)
+        ;; Fallback if there's no exact match
+        (first matches)))))

--- a/src/metabase/sync/sync_metadata/fields/common.clj
+++ b/src/metabase/sync/sync_metadata/fields/common.clj
@@ -68,16 +68,14 @@
                     (when (= (canonical-name field-metadata)
                              (canonical-name other-field-metadata))
                       other-field-metadata))
-                  other-metadata)
-        num-matches (count matches)]
-    (cond
-      (zero? num-matches)
+                  other-metadata)]
+    (case (count matches)
+      0
       nil
 
-      (= 1 num-matches)
+      1
       (first matches)
 
-      :else
       (or
         (some (fn [match]
                 (when (= (:name field-metadata) (:name match))

--- a/src/metabase/sync/sync_metadata/fields/common.clj
+++ b/src/metabase/sync/sync_metadata/fields/common.clj
@@ -2,10 +2,12 @@
   "Schemas and functions shared by different `metabase.sync.sync-metadata.fields.*` namespaces."
   (:require
    [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.shared.util.i18n :as i18n]
    [metabase.sync.interface :as i]
    [metabase.sync.util :as sync-util]
    [metabase.util :as u]
    [metabase.util.i18n :refer [trs]]
+   [metabase.util.log :as log]
    [metabase.util.malli :as mu]
    [metabase.util.malli.registry :as mr]
    [metabase.util.malli.schema :as ms]))
@@ -76,10 +78,11 @@
       1
       (first matches)
 
-      (or
-        (some (fn [match]
-                (when (= (:name field-metadata) (:name match))
-                  match))
-              matches)
-        ;; Fallback if there's no exact match
-        (first matches)))))
+      (if-let [exact (some (fn [match]
+                             (when (= (:name field-metadata) (:name match))
+                               match))
+                           matches)]
+        exact
+        (do
+          (log/warn (i18n/trs "Found multiple matching field metadata for:") (:name field-metadata) (map :name matches))
+          (first matches))))))

--- a/test/metabase/sync/sync_metadata/fields_test.clj
+++ b/test/metabase/sync/sync_metadata/fields_test.clj
@@ -187,3 +187,25 @@
                   :semantic-type     :type/FK
                   :fk-target-exists? true}
                  (state))))))))
+
+(deftest case-sensitive-conflict-test
+  (testing "Two columns with same lower-case name can be synced (#17387)"
+    (one-off-dbs/with-blank-db
+      (doseq [statement [;; H2 needs that 'guest' user for QP purposes. Set that up
+                         "CREATE USER IF NOT EXISTS GUEST PASSWORD 'guest';"
+                         ;; Keep DB open until we say otherwise :)
+                         "SET DB_CLOSE_DELAY -1;"
+                         ;; create table & load data
+                         "DROP TABLE IF EXISTS \"birds\";"
+                         "CREATE TABLE \"birds\" (\"event\" VARCHAR, \"eVent\" VARCHAR);"
+                         "GRANT ALL ON \"birds\" TO GUEST;"
+                         (str "INSERT INTO \"birds\" (\"event\", \"eVent\") VALUES "
+                              "('a', 'b'),  "
+                              "('c', 'd');")]]
+        (jdbc/execute! one-off-dbs/*conn* [statement]))
+      (let [sync-info (sync/sync-database! (mt/db))
+            field-sync-info (->> sync-info
+                                 (m/find-first (comp #{"metadata"} :name))
+                                 :steps
+                                 (m/find-first (comp #{"sync-fields"} first)))]
+        (is (=? ["sync-fields" {:total-fields 2 :updated-fields 2}] field-sync-info))))))

